Add laser-phasechange example (#475)
lethe-cfd/lethe: doc/source/examples/multiphysics/laser-melting/laser-melting.rst

==========================
Laser melting
==========================

----------------------------------
Features
----------------------------------
- Solver: ``gls_navier_stokes_2d`` 
- Laser heat source
- Phase change (solid-liquid)
- Convection-radiation heat transfer boundary condition
- Mesh adaptation using temperature

This example simulates a two-dimensional melt pool with a laser. 

A stainless steel rectangular block melts using a laser beam that emits perpendicular to the top surface of the block. The laser beam speed is 0.5 m/s. Due to the laser heat source, the stainless steel solid block melts in the direction of the laser. The corresponding parameter file is 
``laser_melting.prm``.

The following schematic describes the geometry and dimensions of the simulation in the :math:`(x,y)` plane:

.. image:: images/laser_phasechange.png
    :alt: Schematic
    :align: center
    :width: 400

------------------------
Location of the example
------------------------
``examples/multiphysics/laser-melting/laser-melting.prm``

--------------
Parameter file
--------------

Time integration is handled by a 1st order backward differentiation scheme 

> `(bdf1)`, for a :math:`0.0035` s simulation time with a constant

Author: applied.

        subsection bc 2
        	set id = 2
    	set type	      = convection-radiation
            set h	              = 15
            set Tinf	      = 20
            set emissivity        = 0.5
        end
        subsection bc 3
        	set id = 3
    	set type	      = convection-radiation
            set h	              = 15
            set Tinf	      = 20
            set emissivity        = 0.5
        end
    end

The ``multiphysics`` subsection enables to turn on `(true)` 
and off `(false)` the physics of interest. Here ``heat transfer`` is enabled.

.. code-block:: text

    #---------------------------------------------------
    # Multiphysics
    #---------------------------------------------------
    subsection multiphysics
        set heat transfer  		= true
    end 
    

In the ``laser parameters`` section, the parameters of the laser model are defined. The exponential decaying model [1] is used to simulate the laser heat source. In the exponential decaying model, the laser heat flux is calculated using the following equation:

    .. math:: 
        q(x,y,z) = \frac{\eta \alpha P}{\pi r^2 \mu} \exp{(-\eta \frac{r^2}{R^2})} \exp{(- \frac{|z|}{\mu})}

> where :math:`\eta`, :math:`\alpha`, :math:`P`, :math:`R`, :math:`\mu`, :math:`r`, \:math:`z` denote concentration factor, absorptivity, laser power, beam radius, penetration depth, radial distance from the laser focal point, and axial distance from the laser focal point, respectively.

Reviewer: I would add a bit of an explanation also as to what this equation means. This is not that easy to understand from the get go.

> where :math:`\eta`, :math:`\alpha`, :math:`P`, :math:`R`, :math:`\mu`, :math:`r`, \:math:`z` denote concentration factor, absorptivity, laser power, beam radius, penetration depth, radial distance from the laser focal point, and axial distance from the laser focal point, respectively.

Author: I've explained the laser parameters in the parameters section in detail. I just mentioned the formula here as a reminder.

> where :math:`\eta`, :math:`\alpha`, :math:`P`, :math:`R`, :math:`\mu`, :math:`r`, \:math:`z` denote concentration factor, absorptivity, laser power, beam radius, penetration depth, radial distance from the laser focal point, and axial distance from the laser focal point, respectively.

Author: done.

.. note:: 
    The scanning path of the laser is defined using a Function exptression in the ``path`` subsection.

.. code-block:: text

    #---------------------------------------------------
    # Laser parameters
    #---------------------------------------------------
    subsection laser parameters
    	set enable = true
        	set concentration factor 	= 2
        	set power 			= 100
        	set absorptivity 		= 0.6
        	set penetration depth 		= 0.000070
        	set beam radius 		= 0.000050
        	set start time 			= 0
        	set end time 			= 0.0034
        	set beam orientation 		= y
        	subsection path
        		set Function expression =  0.5 * t; 0.000500
        	end
    end    

> The laser heat source locally melts the material, which is initially in the solid phase according to the definition of the ``solidus temperature``. Hence, the physical properties should be defined using ``phase_change`` models. In the ``physical properties`` subsection, the physical properties of the different phases of the fluid are defined:

Reviewer: Maybe point out to the Stefan example for a more thorough explanation of this?

Author: done.

``set initial refinement=8``. 

.. code-block:: text
        
    #---------------------------------------------------
    # Mesh
    #---------------------------------------------------
    subsection mesh
            set type = dealii
            set grid type = subdivided_hyper_rectangle
            set grid arguments 	= 8, 2 : -0.0001, 0 : 0.0019, 0.000500 : true
            set initial refinement 	= 8
    end
    
In the ``mesh adaptation subsection``, adaptive mesh refinement is 
defined for ``temperature``. ``min refinement level`` and ``max refinement level`` are 4 and 8, respectively.

.. code-block:: text

    #---------------------------------------------------
    # Mesh Adaptation
    #---------------------------------------------------
    subsection mesh adaptation
        set type                    = kelly
        set variable                = temperature
        set fraction type           = fraction
        set max refinement level    = 8
        set min refinement level    = 4
        set frequency               = 1
        set fraction refinement     = 0.5

>         set fraction coarsening     = 0.2

Reviewer: Can you add a justification for this choice?

    end

Call the gls_navier_stokes_2d by invoking:  

``mpirun -np 8 gls_navier_stokes_2d laser_melting.prm``

to run the simulation using eight CPU cores. Feel free to use more.

.. warning:: 
    Make sure to compile lethe in `Release` mode and 
    run in parallel using mpirun. This simulation takes
    :math:`\approx` 5 mins on 8 processes.

-------
Results
-------

> The following image shows the temperature distribution in the simulation domain after 0.0034 seconds.

Reviewer: It would be interesting to add a bit more of a discussion here. Like analyse the results or something.

.. image:: images/temperature.png
    :alt: temperature
    :align: center

>     :width: 800

Reviewer 2: I love the color palette!!!

.. code-block:: text

    #---------------------------------------------------
    # Mesh Adaptation
    #---------------------------------------------------
    subsection mesh adaptation
        set type                    = kelly
        set variable                = temperature
        set fraction type           = fraction
        set max refinement level    = 8
        set min refinement level    = 4
        set frequency               = 1
        set fraction refinement     = 0.5
        set fraction coarsening     = 0.2
    end

----------------------
Running the simulation
----------------------

Call the gls_navier_stokes_2d by invoking:  

``mpirun -np 8 gls_navier_stokes_2d laser_melting.prm``

to run the simulation using eight CPU cores. Feel free to use more.

.. warning:: 
    Make sure to compile lethe in `Release` mode and 
    run in parallel using mpirun. This simulation takes
    :math:`\approx` 5 mins on 8 processes.

-------
Results
-------

The following image shows the temperature distribution in the simulation domain after 0.0034 seconds.

.. image:: images/temperature.png
    :alt: temperature
    :align: center
    :width: 800

> Using ParaView post-processing tools (using a temperature threshold above the liquidus temperature, we can monitor the melted region. The following image shows the melted region at :math:`t=0.00005` s, :math:`0.0005` s, and :math:`0.0025` s. Note that the melted region is visualized using yellow color.

Reviewer: ```suggestion
Using Paraview, we can monitor the melted region by using a temperature threshold above the liquidus temperature. The following image shows, in yellow, the melted region at :math:`t=0.00005` s, :math:`0.0005` s, and :math:`0.0025` s. 
```
